Add Prometheus and docker-compose

Iajrdev opened the pull request:
This PR adds Prometheus and Grafana to the project for monitoring purposes.

- Added docker-compose.yml for setting up Prometheus and Grafana.
- Configured Prometheus to scrape metrics from the Go application.
- Updated documentation with setup instructions.

alipourhabibi commented:
I reviewed the PR
I found an error in `grafana/provisioning/dashboards/my_dashboard.json`
It looks like the json is not valid for dashboard.
I think you should remove the following
```diff
 {
-  "dashboard": {
     "id": null,
     "title": "My App Dashboard",
     "tags": [],
@@ -92,7 +91,5 @@
         ]
       }
     ]
-  },
-  "overwrite": false
 }

```
Apart from that all looks good to me.

alipourhabibi commented on grafana/provisioning/dashboards/my_dashboard.json:
Please change this file based on the previous comment i sent.

alipourhabibi commented:
LGMT

armineyvazi commented:
I've evaluated the changes.

Changed IP addresses to service names for better maintainability and readability.

#### Changes:
- **prometheus/prometheus.yml**:
  ```yaml
  static_configs:
    - targets: ['app:8080']
  ```

- **docker-compose.yml**:
  ```yaml
  services:
    app:
      build: .
      container_name: app
      volumes:
        - .:/app
      working_dir: /app
      command: go run main.go
      ports:
        - "8080:8080"
      restart: unless-stopped
  ```

---

armineyvazi commented on prometheus/prometheus.yml:
```diff
@@ -0,0 +1,9 @@
+  - job_name: 'myapp'
+    metrics_path: '/v2/metrics'
+    static_configs:
+      - targets: ['172.17.0.1:8080']
```
```diff
global:
  scrape_interval: 5s

scrape_configs:

  - job_name: 'myapp'
    metrics_path: '/v2/metrics'
    static_configs:
-     - targets: ['172.17.0.1:8080']
+     - targets: ['host.docker.internal:8080']
```

armineyvazi replied:
I found a mistake in prometheus/prometheus.yml. It looks like the target IP is outdated. I think you should change the following

alipourhabibi replied:
As i understand the `172.17.0.1` refers to the IP address of the default gateway on the default bridge network and is not outdated. However the `host.docker.internal` is used in Docker desktop and won't work on Linux systems without additional configuration.
If we want to support Windows and Linux together we can add the following to the docker-compose.yaml for prometheus
```yaml
    extra_hosts:
      - "host.docker.internal:host-gateway"
```
and use `host.docker.internal` as our targets in prometheus/prometheus.yml
What do you think? 
@Iajrdev

armineyvazi replied:
Thank you for completing this. I think this configuration is okay for Linux, macOS, and Windows OS:

[prometheus/prometheus.yml](https://github.com/PwPJ/linkedlist/pull/11/files/77626bd694fede62db808a8e03fc0872f3a42a66#diff-79c19119fd226b897b5aa398c6623a078e9ae00fbd2fe20807e1fbb373fa0294)
```diff
+     - targets: ['host.docker.internal:8080']
```

[docker-compose.yml](https://github.com/PwPJ/linkedlist/pull/11/files/77626bd694fede62db808a8e03fc0872f3a42a66#diff-e45e45baeda1c1e73482975a664062aa56f20c03dd9d64a827aba57775bed0d3)

```diff
+  extra_hosts:
+    - "host.docker.internal:host-gateway"
```

Iajrdev replied:
To clarify, you're suggesting that we should remove `'172.17.0.1:8080'` from the `prometheus.yml` file and replace it with `'host.docker.internal:8080'`?
@alipourhabibi @armineyvazi

armineyvazi replied:
To clarify, the suggestion is to replace `172.17.0.1:8080` with `host.docker.internal:8080` in the `prometheus.yml` file. Additionally, we should add the following line to the `docker-compose.yaml` file to ensure compatibility across both Windows/Mac and Linux environments:

```yaml
extra_hosts:
  - "host.docker.internal:host-gateway"
```

This configuration will make `host.docker.internal` resolvable within Docker containers on all platforms.

@alipourhabibi @Iajrdev

Iajrdev replied:
I have made changes to the files as per your review and pushed them.
Please review them again.

armineyvazi replied:
I am not sure, but I think it should look like this, not added in Grafana:

```diff

  prometheus:
    image: docker.arvancloud.ir/prom/prometheus:latest
    container_name: prometheus
    volumes:
      - ./prometheus/prometheus.yml:/etc/prometheus/prometheus.yml
      - prometheus_data:/prometheus
    ports:
      - "9090:9090"
+  extra_hosts:
+   - "host.docker.internal:host-gateway"
    restart: unless-stopped

```

If you have a Linux system, please test this configuration to ensure it works correctly. The extra_hosts directive ensures that host.docker.internal is resolvable within the Prometheus container. Add the Grafana service if needed, without extra_hosts unless required for Grafana.

@Iajrdev

armineyvazi reviewed (approved):
**Comment to Approve:**

Approved. The configuration is now compatible with Linux, macOS, and Windows. Adding `host.docker.internal:host-gateway` to `extra_hosts` ensures that the Docker container can correctly reference the host machine across different operating systems. 

Thanks for the update!

Iajrdev commented:
- **Added `healthcheck` for Prometheus**:
  - Implemented a health check for the Prometheus service to ensure it is ready before proceeding with other services.
  - Health check uses the command: `["CMD", "curl", "-f", "http://localhost:9090/-/ready"]`.
  
- **Configured `depends_on` for Grafana**:
  - Updated the Grafana service to depend on the Prometheus service being healthy.
  - This guarantees Grafana will only start after Prometheus has successfully passed its health check.
  
- **Added `healthcheck` for Grafana**:
  - Implemented a health check for the Grafana service to monitor its readiness.
  - Health check uses the command: `["CMD", "curl", "-f", "http://localhost:3000/api/health"]`.
  
@alipourhabibi @armineyvazi 
Please review again.

alipourhabibi commented on docker-compose.yml:
```diff
@@ -0,0 +1,42 @@
+    extra_hosts:
+      - "host.docker.internal:host-gateway"
+    healthcheck:
+      test: ["CMD", "curl", "-f", "http://localhost:9090/-/ready"]
```
As i know the `prom/prometheus:latest` does not have the curl command installed.
So the health check will not ever be successful.
I think you should change it to:
```yaml
test: ["CMD", "wget", "--spider", "--quiet", "http://localhost:9090/-/healthy"]
```

alipourhabibi reviewed (approved):
LGTM
